object/slicer/slicer.go

```diff
@@ -0,0 +1,596 @@
+// within the limit, one object is produced. Note that Slicer can write multiple
+// objects, but returns the root object ID only.
+//
+// If current NeoFS epoch is specified via Options.SetCurrentNeoFSEpoch, it is
+// written to the metadata of all resulting objects as a creation epoch.
+//
+// See also NewSession.
+func New(signer neofscrypto.Signer, cnr cid.ID, owner user.ID, w ObjectWriter, opts Options) *Slicer {
+	return &Slicer{
+		signer: signer,
+		cnr:    cnr,
+		owner:  owner,
+		w:      w,
+		opts:   opts,
+	}
+}
+
+// NewSession creates Slicer which generates objects within provided session.
+// NewSession work similar to New with the detail that the session issuer owns
+// the produced objects. Specified session token is written to the metadata of
+// all resulting objects. In this case, the object is considered to be created
+// by a proxy on behalf of the session issuer.
+func NewSession(signer neofscrypto.Signer, cnr cid.ID, token session.Object, w ObjectWriter, opts Options) *Slicer {
+	return &Slicer{
+		signer:       signer,
+		cnr:          cnr,
+		owner:        token.Issuer(),
+		w:            w,
+		opts:         opts,
+		sessionToken: &token,
+	}
+}
+
+// childPayloadSizeLimit returns configured size limit of the child object's
+// payload which defaults to 1MB.
+func (x *Slicer) childPayloadSizeLimit() uint64 {
+	if x.opts.objectPayloadLimit > 0 {
+		return x.opts.objectPayloadLimit
+	}
+	return 1 << 20
+}
+
+// Slice creates new NeoFS object from the input data stream, associates the
+// object with the configured container and writes the object via underlying
+// ObjectWriter. After a successful write, Slice returns an oid.ID which is a
+// unique reference to the object in the container. Slice sets all required
+// calculated fields like payload length, checksum, etc.
+//
+// Slice allows you to specify string key-value pairs to be written to the
+// resulting object's metadata as object attributes. Corresponding argument MUST
+// NOT be empty or have odd length. Keys SHOULD NOT start with system-reserved
+// '__NEOFS__' prefix.
+//
+// See New for details.
+func (x *Slicer) Slice(data io.Reader, attributes ...string) (oid.ID, error) {
+	if len(attributes)%2 != 0 {
+		panic("attributes must be even number of strings")
+	}
+
+	var rootID oid.ID
+
+	fillCommonMetadata := func(obj *object.Object) {
+		currentVersion := version.Current()
+		obj.SetVersion(&currentVersion)
+		obj.SetContainerID(x.cnr)
+		obj.SetCreationEpoch(x.opts.currentNeoFSEpoch)
+		obj.SetType(object.TypeRegular)
+		obj.SetOwnerID(&x.owner)
+		obj.SetSessionToken(x.sessionToken)
+	}
+
+	var rootHeader object.Object
+	fillCommonMetadata(&rootHeader)
+
+	var rootMeta dynamicObjectMetadata
+	rootMeta.reset()
+
+	bChunk := make([]byte, x.childPayloadSizeLimit()+1)
+
+	n, err := data.Read(bChunk)
+	if err != nil && err != io.EOF {
+		return rootID, fmt.Errorf("read payload chunk: %w", err)
+	}
+
+	fin := n < len(bChunk)
```

Comment:
Why this way to check instead of `io.EOF`? We probably could have `bChunk` of `x.childPayloadSizeLimit()` size then as well. And we can have `n < len(bChunk)` with `err == nil` (and more data to read).

Reply:
nice catch, that's because I wanted to use `ReadFull`, fixed now, check

Reply:
`fin := errors.Is(err, io.EOF)`, isn't it?

Reply:
it can be `io.ErrUnexpectedEOF` because we try to realize does data require slicing or not. So, no

Reply:
discussed in https://github.com/nspcc-dev/neofs-sdk-go/pull/382#discussion_r1177748392